wolfSSL/wolfsentry
# 20230815-docs-and-fixes

* `doc/freertos-lwip-app.md`: added, "Building and Initializing wolfSentry for an application on FreeRTOS/lwIP"

`src/wolfsentry_util.c`:
* add `wolfsentry_cleanup_push()`, `wolfsentry_cleanup_pop()`, and `wolfsentry_cleanup_all()`, for automatic cleanup for integrations and plugins at time of `wolfsentry_shutdown()`;
* semaphore facility abstraction: add `builtin_sem_methods` gated on !`NO_BUILTIN_SEM_API`, and shim in a new `static const struct wolfsentry_semcbs *semcbs` using macros;
* in `wolfsentry_eventconfig_check()`, don't allow `route_private_data_alignment` larger than `route_private_data_size`;
* in `wolfsentry_eventconfig_get_1()`, correctly subtract `route_private_data_padding` from internal `route_private_data_size` to compute exported size;
* in `wolfsentry_context_free()`, add call to `wolfsentry_cleanup_all()`.

`src/lwip/packet_filter_glue.c`:
* add `wolfsentry_cleanup_lwip_filter_callbacks`;
* add `wolfsentry_cleanup_push(..., wolfsentry_cleanup_lwip_filter_callbacks, ...)` calls to all callback installation routines;
* fix callback installation routines to take `WOLFSENTRY_CONTEXT_ARGS_IN`, not ad hoc `struct wolfsentry_context *wolfsentry`.

`src/routes.c`:
* add `clamp_wildcard_fields_to_zero()`, pulled from `wolfsentry_route_insert_1()`;
* call `clamp_wildcard_fields_to_zero()` from `wolfsentry_route_lookup_1()` to assure that a route can be deleted with the same input args with which it was inserted, regardless of wildcards.

`wolfsentry/wolfsentry_errcodes.h`: add `WOLFSENTRY_SUCCESS_ID_ALREADY_OK`;

`tests/unittests.c`:
* harmonize `activate_wolfsentry_lwip()` and `shutdown_wolfsentry_lwip()` with `doc/freertos-lwip-app.md`;
* systematize use of `PRIVATE_DATA_SIZE` and `PRIVATE_DATA_ALIGNMENT`, allowing override;
* add `WOLFSENTRY_TEST_HPI_POSIX_VANILLA` code path to test user-supplied `wolfsentry_host_platform_interface`;
* in `test_static_routes()`, check `clamp_wildcard_fields_to_zero()` in a `wolfsentry_route_delete()` call.

`wolfsentry/wolfsentry.h`: fix `sem_timedwait_cb_t` (add missing `const` attr).

`wolfsentry/wolfsentry_settings.h`:
* add `DO_NOTHING` macro definition;
* add `static const uint32_t __wolfsentry_config` for portable compile-time const, for ease setting up `static const struct wolfsentry_host_platform_interface`.

`Makefile.analyzers`:
* update `CLANG_PATH` with clang-18;
* add `linux-lwip-sanitize-all-gcc` target and add it to `analyze-all`;
* add `linux-lwip-c89-m32-test` target and add it to `check-extra`;
* add missing `test` target to `short-enums-test`;
* add `small-PRIVATE_DATA_ALIGNMENT-test`, `big-PRIVATE_DATA_ALIGNMENT-test`, and `HPI_POSIX_VANILLA-test` targets and add them to `check-extra`.

`scripts/build_wolfsentry_options_h.awk`: fix bug such that input `-Dfoo=0` was detected as a no-value definition due to `awk` treating the zero as false.

`ports/Linux-lwIP/include/lwipopts.h`: reformat C++ comments with C syntax.
